src/cargo/ops/cargo_compile.rs

```diff
@@ -1291,3 +1344,90 @@ fn traverse_and_share(
     new_graph.entry(new_unit.clone()).or_insert(new_deps);
     new_unit
 }
+
+/// Build `glob::Pattern` with informative context.
+fn build_glob(pat: &str) -> CargoResult<glob::Pattern> {
+    glob::Pattern::new(pat).with_context(|| format!("Cannot build glob pattern from `{}`", pat))
+}
+
+/// Emits "package not found" error.
+///
+/// > This function should be used only in package selection processes such like
+/// `Packages::to_package_id_specs` and `Packages::get_packages`.
+fn emit_package_not_found(
+    ws: &Workspace<'_>,
+    opt_names: BTreeSet<&str>,
+    opt_out: bool,
+) -> CargoResult<()> {
+    if !opt_names.is_empty() {
+        anyhow::bail!(
+            "{}package(s) {} not found in workspace `{}`",
+            if opt_out { "excluded " } else { "" },
+            opt_names
+                .iter()
+                .map(|x| x.as_ref())
```

Comment:
I think this can be simplified a bit.

```suggestion
                .into_iter()
```